Recognize `@everyone` as mention in messages

0x-r4bbit opened the pull request:
This "id" or "tag" will be interpeted/interpolated as `@everyone` by
clients.

For this change to work

https://github.com/status-im/markdown/pull/5

is needed.

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (9)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | b62017e5 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3026/1/) | 2022-12-15 11:06:22 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221215-110152-b62017-pr3026.aar) |
| :heavy_check_mark: | b62017e5 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3026/1/) | 2022-12-15 11:06:27 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-221215-110152-b62017-pr3026.zip) |
| :heavy_check_mark: | b62017e5 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3026/1/) | 2022-12-15 11:06:32 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221215-110152-b62017-pr3026.zip) |
| | | | | | | |
| :x: | 7f210fbb | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3026/2/) | 2022-12-19 14:52:31 | ~22 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3026/2/consoleText) |
| :heavy_check_mark: | 7f210fbb | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3026/2/) | 2022-12-19 14:53:36 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221219-145213-7f210f-pr3026.zip) |
| :heavy_check_mark: | 7f210fbb | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3026/2/) | 2022-12-19 14:57:12 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221219-145209-7f210f-pr3026.aar) |
| | | | | | | |
| :x: | c56b364f | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3026/3/) | 2022-12-19 16:04:15 | ~55 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3026/3/consoleText) |
| :heavy_check_mark: | c56b364f | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3026/3/) | 2022-12-19 16:04:30 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221219-160318-c56b36-pr3026.zip) |
| :heavy_check_mark: | c56b364f | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3026/3/) | 2022-12-19 16:06:38 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221219-160318-c56b36-pr3026.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 6598b114 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3026/4/) | 2022-12-20 10:00:26 | ~15 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3026/4/consoleText) |
| :x: | 6598b114 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3026/4/) | 2022-12-20 10:00:34 | ~22 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3026/4/consoleText) |
| :heavy_check_mark: | 6598b114 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3026/4/) | 2022-12-20 10:04:22 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221220-100009-6598b1-pr3026.aar) |
| | | | | | | |
| :x: | 50db05ca | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3026/5/) | 2022-12-20 10:42:40 | ~13 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3026/5/consoleText) |
| :heavy_check_mark: | 50db05ca | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3026/5/) | 2022-12-20 10:43:49 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-221220-104226-50db05-pr3026.zip) |
| :heavy_check_mark: | 50db05ca | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3026/5/) | 2022-12-20 10:46:29 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-221220-104225-50db05-pr3026.aar) |

jrainville commented on protocol/common/message.go:
```diff
@@ -504,7 +529,7 @@ func (m *Message) PrepareContent(identity string) error {
 	// an identity
 	if !m.Mentioned {
-		m.Mentioned = visitor.mentioned
+		m.Mentioned = visitor.mentioned || containsMentionOf(m.Text, EveryoneMentionLiteral)
```
Maybe there is a way to put `@everyone` in the checks done for `runMentionsAndLinksVisitor` but I have no idea how it works so it can be done later.

0x-r4bbit replied:
Ah, that's what I did initially. The problem then is however, that when a node of the AST is visited, we don't know it's surroundings. These will end up in the `default` case (they aren't `ast.Mention` or `ast.Link`).

So the `literal` to inspect will be "@everyone" even though the message might be  [backtick]@everyone[backtick] at which point we don't want this to become recgonized as mention.

The options are to

1. either extend the markdown parser to know what an `@everyone` mention is and that it's to interpret as a mention or
2. Simply check the literal on the entire text (instead of AST)

I went with 2) as it seem simpler

0x-r4bbit replied:
Well, now that I read https://github.com/status-im/status-go/pull/3026#issuecomment-1353427503 I guess I'll have to make it work with 1)  haha

cammellos commented on protocol/common/message.go:
```diff
@@ -70,6 +70,8 @@ const (
 	ContactVerificationStateCanceled
 )
 
+const EveryoneMentionLiteral = "@everyone"
```
You can use lowercase if not exported

0x-r4bbit replied:
Will change this

jrainville commented on protocol/common/message.go:
```diff
@@ -481,6 +483,29 @@ func (v *MentionsAndLinksVisitor) Visit(node ast.Node, entering bool) ast.WalkSt
+
+		if literalLen >= mentionLen+2 && ((idx > 0 && string(literal[idx-1]) == " ") && (string(literal[idx+mentionLen]) == " ")) {
+			return true
+		}
```
Maybe a regex would be better here for the optional spaces

cammellos replied:
we should not use regexes for parsing messages as much as possible, as they tend to lead to performance issue (and possible dos), the markdown library we use tries (I believe we do), to pass only once through the string.

cammellos commented:
We don't do parsing in status-go (at least we should not), we have our markdown library and that's the place where we want to do this for performance reasons, this is also scanning the string multiple times (`Contains`, `IndexOf`, all of this can be easily done in one pass of the string), and also no tests.

Definitely something that should have been flagged @richard-ramos @jonathanbarker please be more careful next time before approving

richard-ramos commented:
Oops! good catch @cammellos!  removing my approval immediatly :-)

jrainville commented:
Sorry about that, I didn't know about the no parsing in status-go.
Also, sorry to JB for being pinged for no reason :joy:

0x-r4bbit commented:
@cammellos thanks for pointing this out. Will move this into markdown's parser

cammellos commented:
> Sorry about that, I didn't know about the no parsing in status-go. Also, sorry to JB for being pinged for no reason joy

No worries :)

>  @cammellos thanks for pointing this out. Will move this into markdown's parser


Thanks,

We kind of need to decided as well how to handle these tags, since we might have custom ones etc, potentially we could even have `@{id}` for example, where everyone is just `@0` or `@0x001` say, and then the might be contextual to the place you receive in, say we add groups eventually (in community `y`, `@0x002` will refer to group 2, which is created by the admin, which translates to `Desktop` for example), or something like that, not sure what's the best

0x-r4bbit commented:
@cammellos what you're describing there are some very interesting ideas. I didn't think about it that way.

It probably makes sense to introduce "namespaces" for different things. For example:

1. `@0x000 - `@0x010`
For system related mentions (like `@everyone`, `@here`, `@channel` etc)

2. `@0x100` - `0x1xx`
For groups. Although this would only allow for 100 groups (which is probably enough), we could make it more explicit and have `@0xg000 - @0xgxx...`

We should nail these down before implementing them. At least the system related namespace

0x-r4bbit commented:
@cammellos added support for recognizing `@0xXXXXX` as per out discussion in: https://github.com/status-im/markdown/pull/5

Will update this PR accordingly

0x-r4bbit committed:
feat: recgonize `0x00001` as system mention

This "id" or "tag" will be interpeted/interpolated as `@everyone` by
clients.

For this change to work

https://github.com/status-im/markdown/pull/5

is needed.

0x-r4bbit commented:
This is green, just not compiling on macOS because something's wrong with CI's macOS
